### PR TITLE
Improve Rebalancing statistics

### PR DESCRIPTION
**Pull request by mrtracy**
Previous refactoring add a negative impact on rebalancing statistics;
specifically, existing replicas of a range were not included in the statistics
when attempting to find a rebalance candidate, which meant that very few
rebalancing opportunities were selected.  This has been fixed by this commit.

Additionally, this commit increases the rate at which rebalances happen; the
previous rate of 5% was proving unresponsive in tests.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/2942)

### CODE REVIEW

File: storage/allocator.go [hunk 1]
```diff
@@ -169,17 +169,17 @@ func (a *Allocator) ComputeAction(zone config.ZoneConfig, desc *roachpb.RangeDes
 // set of stores being considered.
 func (a *Allocator) AllocateTarget(required roachpb.Attributes, existing []roachpb.ReplicaDescriptor, relaxConstraints bool,
 	filter func(storeDesc *roachpb.StoreDescriptor, count, used *stat) bool) (*roachpb.StoreDescriptor, error) {
-	existingNodes := make([]roachpb.NodeID, 0, len(existing))
+	existingNodes := make(nodeIDSet)
```
Comment (tamird) [inline storage/allocator.go]: `make(nodeIDSet, len(existing))`

File: storage/allocator.go [hunk 2]
```diff
@@ -243,13 +243,13 @@ func (a Allocator) RebalanceTarget(storeID roachpb.StoreID, required roachpb.Att
 	if !a.options.AllowRebalance {
 		return nil
 	}
-	existingNodes := make([]roachpb.NodeID, 0, len(existing))
+	existingNodes := make(nodeIDSet)
```
Comment (tamird) [inline storage/allocator.go]: ditto

File: storage/balancer.go [hunk 3]
```diff
@@ -23,22 +23,34 @@ import (
 	"github.com/cockroachdb/cockroach/roachpb"
 )
 
+type nodeIDSet map[roachpb.NodeID]struct{}
+
+func makeNodeIDSet(nodeIDs ...roachpb.NodeID) nodeIDSet {
+	nodeSet := make(nodeIDSet)
```
Comment (tamird) [inline storage/balancer.go]: ditto

**Comment by BramGruneir** [comment]
LGTM.  Do you want to try testing out different levels of the rebalance percentage with the simulator?

**Comment by tamird** [comment]
lgtm

File: storage/balancer.go (hunk 3 above)
Comment (tschottdorf) [inline storage/balancer.go]: what does that even do for a map?

File: storage/balancer.go [hunk 4]
```diff
@@ -23,22 +23,34 @@ import (
...
+	selectGood(sl StoreList, excluded nodeIDSet) *roachpb.StoreDescriptor
 	// selectBad attempts to select a store from the supplied store list that it
 	// considers to be 'Bad' relative to the other stores in the list. Returns
 	// the selected store or nil if no such store can be found.
 	selectBad(sl StoreList) *roachpb.StoreDescriptor
 	// improve attempts to select an improvement over the given store from the
-	// stores in the given store list. Returns the selected store nil if no such
-	// store can be found.
-	improve(store *roachpb.StoreDescriptor, sl StoreList) *roachpb.StoreDescriptor
+	// stores in the given store list. Any nodes in the supplied 'xclude' list
```
Comment (tschottdorf) [inline storage/balancer.go]: s/'x/'ex/

File: storage/balancer.go (hunk 3 above)
Comment (bdarnell) [inline storage/balancer.go]: Sets the initial size of the hash table for a slight optimization.

File: storage/allocator_test.go [hunk 5]
```diff
@@ -1158,55 +1158,55 @@ func Example_rebalancing() {
...
+	// 938 950 999 961 938 921 961 915 946 963 958 967 943 976 971 946 970 966 947 952
+	// 938 965 999 987 922 915 963 921 959 970 973 950 942 994 960 950 992 968 940 951
+	// 928 957 959 957 916 905 969 921 955 981 969 927 939 963 940 948 999 962 951 937
+	// 938 955 945 951 922 892 972 905 952 970 964 945 947 959 945 942 999 967 959 946
+	// 935 954 928 948 923 890 965 893 942 971 958 936 952 949 950 940 999 965 943 935
+	// 951 974 953 971 931 915 991 954 964 999 981 961 983 978 971 966 995 984 981 971
+	// 962 974 946 976 924 912 999 949 958 997 965 964 990 976 965 962 993 965 974 959
+	// 970 962 953 964 929 915 975 949 955 999 962 963 982 985 948 968 974 965 972 934
+	// 974 964 935 947 923 902 964 938 957 999 963 968 968 979 958 965 963 952 967 937
+	// 957 971 936 947 933 905 962 932 951 999 949 968 966 978 962 970 951 944 963 934
```
Comment (bdarnell) [inline storage/allocator_test.go]: What should we be looking for in this test? The the output often changes completely and we seem to just rubber-stamp whatever the new output becomes. Is there a better way to do this test so the output isn't so volatile?

**Comment by bdarnell** [comment]
LGTM

File: storage/allocator_test.go (hunk 5 above)
Comment (mrtracy) [inline storage/allocator_test.go]: I think the target is that, after all of the generations, the difference in size between the various stores should be bounded to a certain level.

Comment (bdarnell) [inline storage/allocator_test.go]: Could we have a test for that then? Something like "variance < X after Y iterations"? It's unclear from this diff whether things are converging any faster than they were before, and none of the other tests are sensitive to the effects of this change.

**Pull request merged**